### ISSUE TITLE
Error: Incomparable types (visitArray_Typevar not overriden in DefaultTypeHierarchy)

### ISSUE THREAD

**Issue by panacekcz** [typetools/checker-framework#1431]
This error seems to happen because the method `visitArray_Typevar` is not overriden in `DefaultTypeHierarchy`. The default implementation inherited from `AbstractAtmComboVisitor` raises the error.

Test case:
```java
class Outer<V>{
	class Inner<T extends V>  {
	}
}
public class IncomparableCrash {
	static void m(Outer<Object>.Inner<int[]> ic) {
	}
}
```
Run the checker at master (3fa3c03ed1bd53124a4389f28b7a18f56533e700). Value checker is used just as an example, happens with other checkers as well:
```
checker/bin/javac -AprintErrorStack -processor org.checkerframework.common.value.ValueChecker IncomparableCrash.java
```
Error output:
```
error: Incomparable types ( @UnknownVal int @UnknownVal [], V extends @UnknownVal Object)visitHistory = VisitHistory(  )
  Compilation unit: IncomparableCrash.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:727)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:25)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:53)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitArray_Typevar(AbstractAtmComboVisitor.java:107)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:292)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:246)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:181)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:159)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:1814)
  org.checkerframework.common.basetype.BaseTypeVisitor.checkTypeArguments(BaseTypeVisitor.java:1990)
  org.checkerframework.common.basetype.BaseTypeValidator.visitParameterizedType(BaseTypeValidator.java:339)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:170)
  org.checkerframework.common.basetype.BaseTypeValidator.visitDeclared(BaseTypeValidator.java:39)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.accept(AnnotatedTypeMirror.java:900)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.scan(AnnotatedTypeScanner.java:89)
  org.checkerframework.framework.type.visitor.AnnotatedTypeScanner.visit(AnnotatedTypeScanner.java:79)
  org.checkerframework.common.basetype.BaseTypeValidator.isValid(BaseTypeValidator.java:69)
  org.checkerframework.common.basetype.BaseTypeVisitor.validateType(BaseTypeVisitor.java:3278)
  org.checkerframework.common.basetype.BaseTypeVisitor.validateTypeOf(BaseTypeVisitor.java:3258)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitVariable(BaseTypeVisitor.java:836)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitVariable(BaseTypeVisitor.java:160)
  com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:864)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:262)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:160)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:91)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:99)
  com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:141)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:568)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:160)
  com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:800)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:68)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:262)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:160)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:81)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:91)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:99)
  com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:133)
  org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:333)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:294)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:160)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:66)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:962)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:494)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:185)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
```

**Comment by panacekcz** [comment]
Similar error can be triggered in `visitUnion_Wildcard`, but this is very artificial code:
```java
import java.io.IOException;
import java.util.ArrayList;
public class IncomparableCrash2 {
	static void m() {
		ArrayList<? super Exception> l=new ArrayList<>();
		try {
			throw new IOException();
		}
		catch(RuntimeException | IOException e) {
			l.add(e);
		}
	}
}
```
Error ouptut (shortened):
```
error: Incomparable types ( @UnknownVal RuntimeException | @UnknownVal IOException, ? extends @UnknownVal Object super @UnknownVal Exception)visitHistory = VisitHistory(  )
  Compilation unit: IncomparableCrash2.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:727)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:25)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:53)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitUnion_Wildcard(AbstractAtmComboVisitor.java:530)
```

**Comment by smillst** [comment]
Thanks for reporting.  Just to clarify (and help prioritize issues), is the first example taken from real code?  If so, what code?

**Comment by panacekcz** [comment]
No, this just came up when debugging another issue and I am not blocked by it in any way.

**Comment by smillst** [comment]
The second crash is completely unrelated to the first and will be fixed separately.  I've opened another issue for it (#1506).
